### ISSUE TITLE
off-strategy: decide whether to do a single-pass compaction based on available space

### ISSUE THREAD

**Issue by denesb** [scylladb/scylladb#16951]
We currently do off-strategy in a single-pass, compacting all sstables in one go, based on the observation that the input sstables (usually coming from repair or streaming) are disjoint, and it is possible to do incremental compaction.
Not all compaction strategies support incremental compaction however, one example being TWCS. If the largest table on a node is a TWCS table, off-strategy can cause problem due to the 100% space amplification it has.

It was proposed to add a configuration item, to force off-strategy to respect `max_threshold` in this case. The configuration item would allow deployments where TWCS tables are not that large, to bypass this and reduce write amplification, caused by multiple compaction passes.

Instead of this config item, we could look at the available space and determine whether a single-pass off-strategy is possible or not. If the available disk space is more than the size of the candidate set (by some margin), then we chose single-pass off-strategy, otherwise we either reduce candidate set size, or choose only `max_threshold` sstables.

@raphaelsc please let me know if I forgot to add something.

**Comment by bhalevy** [comment]
We need to keep in mind that all shards may be running offstartegy compaction in parallel. So the available space needs to be divided by the number of shards to assume the worst case.

Also, if we turn off offstartegy for twcs and temporarily increase min_threshold (e.g to 16), then regular compactiom in the windows will be triggered during repair/streaming and offstrategy compaction wouldn't be needed.  Otherwise, ck-based filtering wouldn't work in the maintenance set, potentially causing temporary increase in read amplification and increased latency due to that, until offstrategy is done.

So all in all, I'm not sure if offstartegy for twcs is really justified (if we increase the mon threshold to prevent too aggressive compaction and too high write amplification).

Also, with tablets this problem should also go away, since we repair the whole tablet as one token range.

The bottom line is that I think we shouldn't invest in this solution unless it's trivial, i.e. has low cost and high gain.

**Comment by raphaelsc** [comment]
> We need to keep in mind that all shards may be running offstartegy compaction in parallel. So the available space needs to be divided by the number of shards to assume the worst case.

Exactly.

> 
> Also, if we turn off offstartegy for twcs and temporarily increase min_threshold (e.g to 16), then regular compactiom in the windows will be triggered during repair/streaming and offstrategy compaction wouldn't be needed. Otherwise, ck-based filtering wouldn't work in the maintenance set, potentially causing temporary increase in read amplification and increased latency due to that, until offstrategy is done.
> 
> So all in all, I'm not sure if offstartegy for twcs is really justified (if we increase the mon threshold to prevent too aggressive compaction and too high write amplification).

It's still good to reduce number of rewrites from 2 to 1. With large data set, it's significant.

> 
> Also, with tablets this problem should also go away, since we repair the whole tablet as one token range.

Exactly.

> 
> The bottom line is that I think we shouldn't invest in this solution unless it's trivial, i.e. has low cost and high gain.

An intermediate solution is pushing offstrategy's space overhead goal as a schema param (a value between 0 and 1), and picking a better default than today's 1.0 (= 100%). I already have that implemented for quick mitigation. And it's already tested by Vlad. If we think implementing this proposal is low cost effort, and therefore worth it, then we can build on top of what I already have and automatically pick a goal value based on available disk space for this shard[1]. Users could still override it by explicitly specifying a value. How does it sound? It of course all depends on the necessity, if having the schema option only is enough until tablet arrives, then we're all good. However, the good advantage of automation based on space is that it will anticipate better than us, it's a proactive solution. With the manual option, it may be too late before we decide to limit space overhead. Tables can grow over time, available space can go short, many variables, and it's hard to track all of that for every user. Disabling off-strategy for TWCS showed some problems in Vlad's tests, like regular compaction not being able to catch up with new files being produced, and fd count exploding. So if we do it, we'd need to implement a back throttler. There's another alternative. Implementing ICS in TWCS, and we kill many birds in one shot. But maybe, it takes longer to implement right. We have to think about how to make TWCS run aware, while not causing a regression on it. It doesn't use partitioned_sstable_set, but rather a specialized one for time series. Maybe, it's possible. 

[1] We could be very conservative. We restrict space overhead in twcs off-strategy to something like 10-20% of available disk space for a given shard. Example, i3en.2xlarge has 8 vcpus, 64G ram, and 5T disk. So each shard, with disk usage of 0%, has (5T - (commitlog_size=ram_size * some_factor=e.g. 2)) / 8 = 609G. If disk is half used, ~300G. So the max allowed space overhead for a shard would be, with 50% disk used, 30-60G. If off-strategy input size is 300G -- which could easily cause node to run out of space -- then we pick space overhead goal of 0.1-0.2, such that usage impact wouldn't exceed 10-20% of available space for a given shard.
When implementing, we also need to remember about making function to return available space sstables::storage specific. With aws s3, I think the space is infinite, unless there's some size quota system.
